torchgeo/models/cnnlstm.py

```diff
@@ -0,0 +1,378 @@
+# Copyright (c) Microsoft Corporation. All rights reserved.
+# Licensed under the MIT License.
+
+"""CNN-LSTM Model Architecture."""
+
+from typing import Any
+
+import timm
+import torch
+import torch.nn as nn
+
+
+class CNN(nn.Module):
+    """CNN backbone using a TIMM model.
+
+    Args:
+        model_name (str): Name of the TIMM model architecture to use.
+        pretrained (bool): Whether to load pretrained weights.
+        output_dim (int, optional): Desired output feature dimension. If None, uses
+            the TIMM model's native feature dimension.
+        freeze_backbone (bool): If True, freezes the backbone parameters.
+        in_chans (int): Number of input channels (e.g., 3 for RGB).
+
+    Raises:
+        ImportError: If TIMM library is not available.
+
+    Forward:
+        x (torch.Tensor): Input tensor of shape either
+            - (B, T, C, H, W) for sequences of images, or
+            - (B, C, H, W) for single images.
+
+    Returns:
+        torch.Tensor: Extracted features of shape (B, T, output_dim) or (B, output_dim).
+    """
+
+    def __init__(
+        self,
+        model_name: str,
+        pretrained: bool = True,
+        output_dim: int | None = None,
+        freeze_backbone: bool = False,
+        in_chans: int = 3,
+    ) -> None:
+        """Initialize CNN backbone with TIMM model.
+
+        Args:
+            model_name: Name of the TIMM model architecture to use.
+            pretrained: Whether to load pretrained weights.
+            output_dim: Desired output feature dimension. If None, uses
+                the TIMM model's native feature dimension.
+            freeze_backbone: If True, freezes the backbone parameters.
+            in_chans: Number of input channels (e.g., 3 for RGB).
+        """
+        super().__init__()
+
+        # Create TIMM model without classification head
+        self.backbone = timm.create_model(
+            model_name, pretrained=pretrained, num_classes=0, in_chans=in_chans
+        )
+
+        # Determine feature dimension
+        with torch.no_grad():
+            dummy_input = torch.randn(1, in_chans, 224, 224)
+            features = self.backbone(dummy_input)
+            feature_dim = features.shape[1]
+
+        # Set output dimension
+        if output_dim is None:
+            output_dim = feature_dim
+        self.output_dim = output_dim
+
+        # Projection layer if needed
+        self.projection: nn.Linear | nn.Identity
+        if feature_dim != output_dim:
+            self.projection = nn.Linear(feature_dim, output_dim)
+        else:
+            self.projection = nn.Identity()
+
+        # Optionally freeze backbone
+        if freeze_backbone:
+            for param in self.backbone.parameters():
+                param.requires_grad = False
```

Comment:
Wondering if this should be here or in the trainer. We do it in the trainer for all other models